File: src/main/python/AoC2020_13.py
#! /usr/bin/env python3
#
# Advent of Code 2020 Day 13
#

import itertools
import sys

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples

TEST1 = """\
939
7,13,x,x,59,x,31,19
"""
TEST2 = """\
0
17,x,13,19
"""
TEST3 = """\
0
67,7,59,61
"""
TEST4 = """\
0
67,x,7,59,61
"""
TEST5 = """\
0
67,7,x,59,61
"""
TEST6 = """\
0
1789,37,47,1889
"""

Input = tuple[int, list[tuple[int, int]]]
Output1 = int
Output2 = int


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        inputs = list(input_data)
        buses = [
            (int(period), offset)
            for offset, period in enumerate(inputs[1].split(","))
            if period != "x"
        ]
        return int(inputs[0]), buses

    def part_1(self, input: Input) -> int:
        target, buses = input
        return next(
            period * cnt
            for cnt in itertools.count(0)
            for period, _ in buses
            if (target + cnt) % period == 0
        )

    def part_2(self, input: Input) -> int:
        _, buses = input
        r = 0
        lcm = buses[0][0]
        for period, offset in buses[1:]:
            while (r + offset) % period != 0:
                r += lcm
            lcm *= period
        return r

    @aoc_samples(
        (
            ("part_1", TEST1, 295),
            ("part_2", TEST1, 1068781),
            ("part_2", TEST2, 3417),
            ("part_2", TEST3, 754018),
            ("part_2", TEST4, 779210),
            ("part_2", TEST5, 1261476),
            ("part_2", TEST6, 1202161486),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2020, 13)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2020_13.java
import static com.github.pareronia.aoc.IterTools.enumerate;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2020_13 extends SolutionBase<AoC2020_13.Notes, Integer, Long> {

    private AoC2020_13(final boolean debug) {
        super(debug);
    }

    public static AoC2020_13 create() {
        return new AoC2020_13(false);
    }

    public static AoC2020_13 createDebug() {
        return new AoC2020_13(true);
    }

    @Override
    protected Notes parseInput(final List<String> inputs) {
        return Notes.fromInput(inputs);
    }

    @Override
    public Integer solvePart1(final Notes notes) {
        return Stream.iterate(0, cnt -> cnt + 1)
            .flatMap(cnt -> notes.buses.stream().map(b -> new int[] { cnt, b.period }))
            .filter(a -> (notes.target + a[0]) % a[1] == 0)
            .map(a -> a[0] * a[1])
            .findFirst().orElseThrow();
    }

    @Override
    public Long solvePart2(final Notes notes) {
        long r = 0;
        long lcm = notes.buses.get(0).period();
        for (int i = 1; i < notes.buses.size(); i++) {
            final Bus bus = notes.buses.get(i);
            while ((r + bus.offset()) % bus.period() != 0) {
                r += lcm;
            }
            lcm *= bus.period();
        }
        return r;
    }

    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "295"),
        @Sample(method = "part2", input = TEST2, expected = "3417"),
        @Sample(method = "part2", input = TEST3, expected = "754018"),
        @Sample(method = "part2", input = TEST4, expected = "779210"),
        @Sample(method = "part2", input = TEST5, expected = "1261476"),
        @Sample(method = "part2", input = TEST6, expected = "1202161486"),
    })
    public static void main(final String[] args) throws Exception {
        AoC2020_13.create().run();
    }

    private static final String TEST1 = """
            939
            7,13,x,x,59,x,31,19
            """;
    private static final String TEST2 = """
            999
            17,x,13,19
            """;
    private static final String TEST3 = """
            999
            67,7,59,61
            """;
    private static final String TEST4 = """
            999
            67,x,7,59,61
            """;
    private static final String TEST5 = """
            999
            67,7,x,59,61
            """;
    private static final String TEST6 = """
            999
            1789,37,47,1889
            """;

    record Bus(int period, int offset) {}

    record Notes(int target, List<Bus> buses) {

        public static Notes fromInput(final List<String> inputs) {
            final int target = Integer.parseInt(inputs.get(0));
            final List<Bus> buses = enumerate(Arrays.stream(inputs.get(1).split(",")))
                .filter(e -> !"x".equals(e.value()))
                .map(e -> new Bus(Integer.parseInt(e.value()), e.index()))
                .toList();
            return new Notes(target, buses);
        }
    }
}